src/others/minimax.js

```diff
@@ -0,0 +1,119 @@
+(function (exports) {
+  'use strict';
+  /* eslint max-params: 0 */
+
+  /**
+   * Minimax (sometimes MinMax, MM[1] or saddle point[2]) is a decision rule used in artificial intelligence,
+   * decision theory, game theory, statistics, and philosophy for minimizing the possible loss for a worst case (maximum loss) scenario.
+   * Optimized with alpha-beta pruning.
+   *  {@link https://en.wikipedia.org/wiki/Minimax}
+   *  {@link https://en.wikipedia.org/wiki/Alpha%E2%80%93beta_pruning}
+   *
+   * @public
+   * @module others/minimax
+   *
+   * @example
+   *
+   * var miniMax =
+   *  require('path-to-algorithms/src/others/minimax').minimax;
+   * var result = minimax(
+   *   [1, 2, 3],
+   *   true,
+   *   5,
+   *   -Infinity,
+   *   Infinity,
+   *   state => ({ move: 0, state: [2, 3, 4] }),
+   *   state => state[1] < 3,
+   *   state => state[1]
+   * );
+   *
+   * @param {*} state Current game state
+   * @param {Boolean} maximize Defines if the result should be maximized or minimized
+   * @param {Number} depth Defines the maximum depth search
+   * @param {Number} alpha Maximum score that the minimizing player is assured
+   * @param {Number} beta Minimum score that the maximizing player is assured
+   * @param {Function} getPossibleNextStatesFn Function which returns all possible next moves with states .
+   * @param {Function} isGameOverFn Function which returns if game is over.
+   * @param {Function} getScoreFn Function which returns score.
+   * @return {{score: Number, move: *}} which contains the minimum coins from the given
+   *  list, required for the change.
+   */
+  function minimax(
+    state,
+    maximize,
+    depth,
+    alpha,
+    beta,
+    getPossibleNextStatesFn,
+    isGameOverFn,
+    getScoreFn
+  ) {
+    if (depth === 0 || isGameOverFn(state)) {
+      const score = getScoreFn(state);
+      return {score, move: null};
+    }
+
+    const possibleMoveResults = getPossibleNextStatesFn(state);
+
+    if (maximize) {
+
+      let maxResult = {score: -Infinity, move: null};
+
+      for (const next of possibleMoveResults) {
+        const result = minimax(
+          next.state,
+          false,
+          depth - 1,
+          alpha,
+          beta,
+          getPossibleNextStatesFn,
+          isGameOverFn,
+          getScoreFn
+        );
+
+        if (result.score > maxResult.score) {
+          maxResult = {score: result.score, move: next.move};
+        }
+
+        alpha = Math.max(alpha, result.score);
+
+        if (alpha >= beta) {
```

Comment:
Is this used for an optimization so we stop the loop earlier?

Reply:
If we stop when `alpha === beta` can't we miss later cases when `alpha` is smaller than `beta`, which will lead to a better outcome?

Reply:
That's right, it's an optimization called "Alpha-beta pruning".
That's not the case. If alpha is larger then beta, then we know that this whole leaf won't be chosen anyway. I couldn't explain it better than 20 seconds of [this video](https://youtu.be/l-hh51ncgDI?t=320).